sys/linux/init_images.go

```diff
@@ -0,0 +1,243 @@
+			newSize = segment.offset.Val + segment.size.Val
+		}
+	}
+	if newSize > imageMaxSize {
+		newSize = imageMaxSize
+	}
+	parsed.size.Val = newSize
+
+	// Drop 0-size segments.
+	parsed.filterSegments(func(i int, segment *mountImageSegment) bool {
+		return segment.size.Val > 0
+	})
+}
+
+type sortSegments struct {
+	parsed *mountImageArgs
+}
+
+func (s sortSegments) Len() int { return len(s.parsed.segments) }
+func (s sortSegments) Swap(i, j int) {
+	inner := s.parsed.segmentsGroup.Inner
+	inner[i], inner[j] = inner[j], inner[i]
+	s.parsed.segments[i], s.parsed.segments[j] = s.parsed.segments[j], s.parsed.segments[i]
+}
+func (s sortSegments) Less(i, j int) bool {
+	segments := s.parsed.segments
+	if segments[i].offset.Val != segments[j].offset.Val {
+		return segments[i].offset.Val < segments[j].offset.Val
+	}
+	return segments[i].size.Val < segments[j].size.Val
+}
+
+func (arch *arch) fixUpSyzMountImage(c *prog.Call) {
+	// Previously we did such a sanitization right in the common_linux.h, but this was problematic
+	// for two reasons:
+	// 1) It further complicates the already complicated executor code.
+	// 2) We'd need to duplicate the logic in Go for raw image extraction.
+	// So now we do all the initialization in Go and let the C code only interpret the commands.
+	ret, err := parseSyzMountImage(c)
+	if err != nil {
+		deactivateSyzMountImage(c)
+		return
+	}
+	const maxImageSegments = 4096
+	ret.filterSegments(func(i int, _ *mountImageSegment) bool {
+		return i < maxImageSegments
+	})
+	fixUpImageSegments(ret)
+}
+
+type mountImageArgs struct {
+	size          *prog.ConstArg
+	segmentsCount *prog.ConstArg
+	segmentsGroup *prog.GroupArg
+	segments      []*mountImageSegment
+}
+
+func (m *mountImageArgs) filterSegments(filter func(int, *mountImageSegment) bool) {
+	newArgs := []prog.Arg{}
+	newSegments := []*mountImageSegment{}
+	for i, segment := range m.segments {
+		if filter(i, segment) {
+			newSegments = append(newSegments, segment)
+			newArgs = append(newArgs, m.segmentsGroup.Inner[i])
+		}
+	}
+	m.segments = newSegments
+	m.segmentsGroup.Inner = newArgs
+	m.segmentsCount.Val = uint64(len(newArgs))
+}
+
+type mountImageSegment struct {
+	data       *prog.DataArg
+	offset     *prog.ConstArg
+	size       *prog.ConstArg
+	parseError error
+}
+
+func parseImageSegment(segmentArg prog.Arg) *mountImageSegment {
+	ret := &mountImageSegment{}
+	segmentFields, ok := segmentArg.(*prog.GroupArg)
+	if segmentFields == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("it is not a group")}
+	}
+	if len(segmentFields.Inner) != 3 {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid number of nested fields")}
+	}
+	dataPtr, ok := segmentFields.Inner[0].(*prog.PointerArg)
+	if dataPtr == nil || dataPtr.Res == nil || !ok {
+		return &mountImageSegment{parseError: fmt.Errorf("invalid data field ptr")}
+	}
+	ret.data, ok = dataPtr.Res.(*prog.DataArg)
+	if ret.data == nil || !ok {
```

Comment:
We won't store a typed nil inside of arguments, so I think all "== nil" checks in this function can be removed.
We may store nil inside of args in some cases, but then the cast to (*prog.DataArg) will fail.